enf/resource_enf_endpoint_ratelimit.go

```diff
@@ -0,0 +1,177 @@
+package enf
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/xaptum/go-enf/enf"
+)
+
+func resourceEnfEndpointRateLimit() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceEnfEndpointRateLimitCreate,
+		Read:   resourceEnfEndpointRateLimitRead,
+		Update: resourceEnfEndpointRateLimitUpdate,
+		Delete: resourceEnfEndpointRateLimitDelete,
+
+		Schema: map[string]*schema.Schema{
+			"endpoint": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"packets_per_second": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"packets_burst_size": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"bytes_per_second": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"bytes_burst_size": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"inherit": {
+				Type:     schema.TypeBool,
+				Required: true,
+			},
+			"type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "current",
+			},
+		},
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+	}
+}
+
+func resourceEnfEndpointRateLimitCreate(d *schema.ResourceData, meta interface{}) error {
+	// The ENF automatically creates a rate limit for each endpoint, so just update the values.
+	return resourceEnfEndpointRateLimitUpdate(d, meta)
+}
+
+func resourceEnfEndpointRateLimitRead(d *schema.ResourceData, meta interface{}) error {
+	switch d.Get("type") {
+	case "current":
+		return endpointCurrentRateLimitRead(d, meta)
+	case "max":
+		return endpointMaxRateLimitRead(d, meta)
+	default:
+		return fmt.Errorf("Invalid rate limit type %s.The rate limit type must be either current or max", d.Get("type"))
+	}
+}
+
+func endpointCurrentRateLimitRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*EnfClient).Client
+	endpointRateLimit, resp, err := client.Endpoint.GetCurrentRateLimits(context.Background(), d.Get("endpoint").(string))
+	if err != nil {
+		if resp.StatusCode == http.StatusNotFound {
```

Comment:
Here too, not sure we need to check for 404 explicitly. That code should never be returned.

Reply:
Would the code be returned in the event that a user specifies an endpoint IP in the Terraform config file that doesn't exist yet?

Reply:
Yes, that's a good point.  It probably is.  Try it and find out.

Would be nice to throw an error saying that the endpoint (network, domain...) Was not valid.

Reply:
So a 400 error is returned when I enter an invalid domain, network, or endpoint, not a 404. In Terraform, this is the error message that is currently displayed:

```
Error: PUT https://dev.xaptum.io/api/xcr/v2/cxns/fd00:880:8000::fd3d:6c94:0/ep_rate_limits/current: [400] generic_error - Unable to update current endpoint rate limits
```

I could go through some work to be more specific with this error message (saying the domain, network, or endpoint was invalid, for instance), but I don't know if it's worth the time right now. Letting all errors be handled with a single error check in the `SetXRateLimits` method seems like a fine approach to me.

@drbild What was the idea with having an invalid IPv6 in the path give a 400, not a 404? Intuitively, if the domain, network, or endpoint doesn't even exist I'd think the server would give me 404, but maybe that's just me.

Reply:
I agree that a 404 would make more sense and be more "correct" by REST standards.

The `xcr` used to return 500 for almost any error, whether in the user-provided request or an actual server error.

We got it fixed to return `4xx` codes for user errors, but I don't think Venkat took it far enough to differentiate bad request, not found, etc.

Reply:
All that to say:
I agree, remove the explicit check for `http.StatusNotFound`.